rbac/management/group/relation_api_dual_write_group_handler.py

```diff
@@ -107,3 +110,58 @@ def _replicate(self):
             )
         except Exception as e:
             raise DualWriteException(e)
+
+    def replicate_added_role(self, role: Role):
+        """Replicate added role. role needs to be locked"""
+
+        if not self.replication_enabled():
+            return
+
+        if role.system:
+            try:
+                mapping = (
+                    BindingMapping.objects.select_for_update()
+                    .filter(
+                        role=role,
+                        resource_type_namespace="rbac",
+                        resource_type_name="workspace",
+                        # TODO: don't use org id once we have workspace built ins
+                        resource_id=self.group.tenant.org_id,
+                    )
+                    .get()
+                )
+                self.group_relations_to_add.append(mapping.add_group_to_bindings(str(self.group.uuid)))
+                mapping.save(force_update=True)
+            except BindingMapping.DoesNotExist:
+                mapping = BindingMapping.for_group_and_role(role, str(self.group.uuid), self.group.tenant.org_id)
+                mapping.save(force_insert=True)
+                self.group_relations_to_add.extend(mapping.as_tuples())
+        else:
+            # Intentionally queried this way to avoid use of cached binding_mappings
+            # There is a risk of phantom reads here, but this is why we lock the custom Role
+            bindings: Iterable[BindingMapping] = BindingMapping.objects.filter(role=role).select_for_update()
+
+            if sum(1 for _ in bindings) == 0:
+                mapping = BindingMapping.for_group_and_role(role, str(self.group.uuid), self.group.tenant.org_id)
+                mapping.save(force_insert=True)
+                self.group_relations_to_add.extend(mapping.as_tuples())
+            else:
+                for mapping in bindings:
+                    self.group_relations_to_add.append(mapping.add_group_to_bindings(str(self.group.uuid)))
+                    mapping.save(force_update=True)
+
+        self._replicate()
+
+    def replicate_removed_role(self, role: Role):
+        if not self.replication_enabled():
+            return
+        # This logic can be the same for system vs custom b/c we'd never create a binding in this path
+        bindings: Iterable[BindingMapping] = role.binding_mappings.select_for_update().all()
+        for binding in bindings:
+            self.group_relations_to_remove.append(binding.remove_group_from_bindings(str(self.group.uuid)))
+            if not binding.is_empty():
```

Comment:
Is it enough condition(only empty group) to remove whole binding mapping  ? Should not we extend it and check whether permissions are empty as well ?
so together:
```
  def is_empty(self):
        return len(self.mappings["groups"]) == 0 && len(self.mappings["permissions"]) == 0 
```

Reply:
If the permissions are empty the role binding should have already been removed by the role update.

However it does beg the question if we should _only_ remove the binding in role updates. We probably should actually, due to the answer I gave above. So I think this empty and removal check is an error – we should **not** remove bindings in this path for **custom roles.**

For **system roles** however, we **can** remove bindings when there are no more groups, because they will be added back when the group is added.